components/authz-service/storage/postgres/migration/sql/62_add_not_null_constraints.up.sql

```diff
@@ -0,0 +1,111 @@
+
+-- we have validation here so nothing should be updated
+DELETE FROM iam_roles WHERE actions IS NULL;
+ALTER TABLE iam_roles ALTER COLUMN actions SET NOT NULL;
+
+DELETE FROM iam_rule_conditions WHERE rule_db_id IS NULL;
+ALTER TABLE iam_rule_conditions ALTER COLUMN rule_db_id SET NOT NULL;
+
+DELETE FROM iam_staged_project_rules WHERE project_id IS NULL;
+ALTER TABLE iam_staged_project_rules ALTER COLUMN project_id SET NOT NULL;
+
+DELETE FROM iam_staged_rule_conditions WHERE rule_db_id IS NULL;
+ALTER TABLE iam_staged_rule_conditions ALTER COLUMN rule_db_id SET NOT NULL;
+
+DELETE FROM iam_statement_projects WHERE statement_id IS NULL;
+ALTER TABLE iam_statement_projects ALTER COLUMN statement_id SET NOT NULL;
+DELETE FROM iam_statement_projects WHERE project_id IS NULL;
+ALTER TABLE iam_statement_projects ALTER COLUMN project_id SET NOT NULL;
+
+-- move code to update policy_change_tracker and notify into a PG function so we can
+-- leverage it from here and the go code.
+CREATE OR REPLACE FUNCTION notify_policy_change ()
+    RETURNS VOID
+    AS $$
+
+    UPDATE policy_change_tracker SET policy_change_id = uuid_generate_v4();
+
+    NOTIFY policychange;
+$$
+LANGUAGE SQL;
+
+CREATE OR REPLACE FUNCTION purge_statements_with_no_projects() RETURNS TRIGGER AS $$
+  BEGIN
+    -- delete all statements that now have no projects
+    DELETE FROM iam_statements s
+      WHERE NOT EXISTS (SELECT 1 FROM iam_statement_projects p WHERE s.db_id = p.statement_id AND p.project_id IS NOT NULL);
+
+    -- clean up any policies that don't have statements as a result
+    DELETE FROM iam_policies p
+        WHERE NOT EXISTS (SELECT 1 FROM iam_statements s WHERE p.db_id = s.policy_id)
+        AND p.type != 'chef-managed';
+
+    PERFORM notify_policy_change();
+
+    RETURN NULL;
+  END
+$$
+LANGUAGE plpgsql;
+
+-- run what is in our cleanup function once
+DELETE FROM iam_statements s
+  WHERE NOT EXISTS (SELECT 1 FROM iam_statement_projects p WHERE s.db_id = p.statement_id AND p.project_id IS NOT NULL);
+
+DELETE FROM iam_policies p
+    WHERE NOT EXISTS (SELECT 1 FROM iam_statements s WHERE p.db_id = s.policy_id)
+    AND p.type != 'chef-managed';
+
+CREATE TRIGGER on_project_deletion AFTER DELETE ON iam_projects
+FOR EACH ROW
+EXECUTE PROCEDURE purge_statements_with_no_projects();
+
+-- ADD notify_policy_change TO OUR ROLE'S TRIGGER! Everything else below is the same as 61 schema migration:
+
+-- i wanted to use the above FKEY constraint with ON DELETE to clean up any statement where the
+-- role getting removed would result in a NULL role_id in iam_statements. that is doable, but
+-- what becomes tricky is adding the additional logic to completely remove the statement if the affected
+-- policy would also have no actions. so instead of splitting that logic up, let's handle it all in a trigger.
+
+CREATE OR REPLACE FUNCTION purge_statements_with_no_actions_or_role() RETURNS TRIGGER AS $$
```

Comment:
this function makes me think we'd purge a statement missing _either_ actions or role . . . maybe `purge_statements_missing_both_actions_and_role` but also, we're not strictly purging here, we're doing other work (deleting the role from statements). 

Could we have the role deletion take place outside of the trigger, then call the trigger _ON DELETE_ of roles?

Reply:
Ah yeah that name isn't great. Good catch! It could take place outside the trigger. Don't see a clear advantage either way. Will update function name for starters.